contracts/Wallet.sol

```diff
@@ -1,68 +1,150 @@
+    function getTokenType(address tokenAddress) private view returns (TokenType) {
+        // Checking the token's interface via ERC165
+        if (tokenAddress.supportsInterface(type(IERC721).interfaceId)) {
+            return TokenType.ERC721;
+        } else if (tokenAddress.supportsInterface(type(IERC1155).interfaceId)) {
+            return TokenType.ERC1155;
+        } else {
+            return TokenType.ERC20; // Defaults to ERC-20 if not ERC-721 or ERC-1155
         }
+    }
 
-        // Fill in ERC-721 token information
-        for (uint256 i = 0; i < erc721Tokens.length; i++) {
-            paginatedInfo[count++] = TokenInfo({
-                tokenAddress: erc721Tokens[i],
-                balance: 0,
-                name: "",
-                symbol: "",
-                decimals: 0
-            });
+    // Private function to get information about a token
+    function getTokenInfo(address tokenAddress, address user) private view returns (TokenInfo[] memory) {
+        TokenType tokenType = getTokenType(tokenAddress);
+        TokenInfo[] memory infos = new TokenInfo[](0);
+        // Handling for ERC20 tokens: Fetches balance, name, symbol, and decimals.
+        // Only includes the token if the user has a non-zero balance
+        if (tokenType == TokenType.ERC20) {
+            IERC20Metadata token = IERC20Metadata(tokenAddress);
+            uint256 balance = token.balanceOf(user);
+            if (balance > 0) {
+                infos = new TokenInfo[](1);
+                infos[0] = TokenInfo({
+                    tokenAddress: tokenAddress,
+                    balance: balance,
+                    name: token.name(),
+                    symbol: token.symbol(),
+                    decimals: token.decimals()
+                });
+            }
+        // Handling for ERC721 tokens: Fetches balance, name, and symbol.
+        } else if (tokenType == TokenType.ERC721) {
+            IERC721Metadata token = IERC721Metadata(tokenAddress);
+            uint256 balance = token.balanceOf(user);
+            if (balance > 0) {
+                infos = new TokenInfo[](1);
+                infos[0] = TokenInfo({
+                    tokenAddress: tokenAddress,
+                    balance: balance,
+                    name: token.name(),
+                    symbol: token.symbol(),
+                    decimals: 0 // Set decimals to 0
+                });
+            }
+        } else if (tokenType == TokenType.ERC1155) {
+            ICRC1155Enumerable tokenEnumerable = ICRC1155Enumerable(tokenAddress);
+            ICRC1155Metadata tokenMetadata = ICRC1155Metadata(tokenAddress);
+            uint256 tokenCount = tokenEnumerable.tokenCountOf(user);
+            infos = new TokenInfo[](tokenCount);
+
+            for (uint256 i = 0; i < tokenCount; i++) {
+                uint256 tokenId = tokenEnumerable.tokenOfOwnerByIndex(user, i);
+                uint256 balance = tokenEnumerable.balanceOf(user, tokenId);
```

Comment:
For ERC1155 case, the `balance` means the number of token ids. This is similar to ERC721.

Reply:
Solved, it's returns the number of distinct token IDs right now.